Changes in module 'nf-core/samtools/dict'
--- modules/nf-core/samtools/dict/main.nf
+++ modules/nf-core/samtools/dict/main.nf
@@ -19,12 +19,13 @@
 
     script:
     def args = task.ext.args ?: ''
+    def dict = fasta.getSimpleName()
     """
     samtools \\
         dict \\
         $args \\
         $fasta \\
-        > ${fasta}.dict
+        > ${dict}.dict
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
