Changes in module 'nf-core/ensemblvep/vep'
--- modules/nf-core/ensemblvep/vep/main.nf
+++ modules/nf-core/ensemblvep/vep/main.nf
@@ -33,10 +33,11 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
     def dir_cache = cache ? "\${PWD}/${cache}" : "/.vep"
     def reference = fasta ? "--fasta $fasta" : ""
+
     """
     vep \\
         -i $vcf \\
-        -o ${prefix}.${file_extension}.gz \\
+        -o ${prefix}.ann.${file_extension}.gz \\
         $args \\
         $compress_cmd \\
         $reference \\
@@ -58,9 +59,9 @@
     stub:
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
-    touch ${prefix}.vcf.gz
-    touch ${prefix}.tab.gz
-    touch ${prefix}.json.gz
+    touch ${prefix}.ann.vcf.gz
+    touch ${prefix}.ann.tab.gz
+    touch ${prefix}.ann.json.gz
     touch ${prefix}.summary.html
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
